[UnifiedPDF] Selection needs to keep tracking the mouse when it's dragged outside of the page area

aprotyas opened the pull request:
#### 9da7290f216c59af48811f6a5e3fad3d468dfbb8
<pre>
[UnifiedPDF] Selection needs to keep tracking the mouse when it&apos;s dragged outside of the page area
<a href="https://bugs.webkit.org/show_bug.cgi?id=269012">https://bugs.webkit.org/show_bug.cgi?id=269012</a>
<a href="https://rdar.apple.com/122574603">rdar://122574603</a>

Reviewed by Darin Adler.

Our selection tracking logic was robust to mouse events outside of the
page area, but we were limited by shortcutting out of our mouse event
handling logic if the received event was not on the page area.

This patch follows the fixes in 276852@main and adopts
PDFDocumentLayout::nearestPageIndexForDocumentPoint to determine which
is the actionable page for a mouse event. This way we don&apos;t early return
from our mouse event handling logic if the event is not on the page
area, and our selection tracking can keep track even when dragged
outside of the page area.

Finally, we make a small refactor in handleMouseEvent to call
stopTrackingSelection as appropriate whenever we leave the method. This
is necessitated by the fact that we may _want_ to clear selection
tracking state, but we return early in certain display modes. This is
explained in the inline comment.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):

Canonical link: <a href="https://commits.webkit.org/277133@main">https://commits.webkit.org/277133@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c51a19c75f7b8a3551d587b95c3ce53cd55d8972

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46686 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25846 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49305 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49363 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42733 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23308 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38054 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22825 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/49305 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19327 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20200 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/49305 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4732 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42960 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/49305 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51224 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21695 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18071 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/51224 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22983 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/49305 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/51224 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10335 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23450 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22688 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2016,28 +2017,33 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
 
+    // Even if the mouse event isn't handled (e.g. because the event is over a page we shouldn't
+    // display in Single Page mode), we should stop tracking selections (and soon autoscrolling) on MouseUp.
+    auto stopSelectionTrackingIfNeeded = makeScopeExit([this, isMouseUp = event.type() == WebEventType::MouseUp] {
+        if (isMouseUp)
+            stopTrackingSelection();
+    });
```
How about this instead?
```
auto stopSelectionTrackingIfNeeded = makeScopeExit(event.type() == WebEventType::MouseUp
    ? [this] { stopTrackingSelection(); } : [this] { });
```

aprotyas replied:
I tried doing this, but the compiler gets confused by the expression and complains that the two operand types (both lambdas) are incompatible. Plus, `check-webkit-style` complains about having two statements in the same line, though that can be fixed with some new line wrapping.

aprotyas replied:
I'll keep it as-is for now!

aprotyas committed:
[UnifiedPDF] Selection needs to keep tracking the mouse when it's dragged outside of the page area
https://bugs.webkit.org/show_bug.cgi?id=269012
rdar://122574603

Reviewed by Darin Adler.

Our selection tracking logic was robust to mouse events outside of the
page area, but we were limited by shortcutting out of our mouse event
handling logic if the received event was not on the page area.

This patch follows the fixes in 276852@main and adopts
PDFDocumentLayout::nearestPageIndexForDocumentPoint to determine which
is the actionable page for a mouse event. This way we don't early return
from our mouse event handling logic if the event is not on the page
area, and our selection tracking can keep track even when dragged
outside of the page area.

Finally, we make a small refactor in handleMouseEvent to call
stopTrackingSelection as appropriate whenever we leave the method. This
is necessitated by the fact that we may _want_ to clear selection
tracking state, but we return early in certain display modes. This is
explained in the inline comment.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):

Canonical link: https://commits.webkit.org/277133@main
